### PR TITLE
[wasm] Add an internal GC lock API 

### PR DESCRIPTION
**Pull request by lambdageek**

### CODE REVIEW

File: src/mono/wasm/runtime/dotnet-legacy.d.ts [hunk 1]
```diff
@@ -302,6 +302,14 @@ type MONOType = {
      * @deprecated Please use getHeapF64
      */
     getF64: (offset: MemOffset) => number;
+    /**
```
Comment (pavelsavara) [inline src/mono/wasm/runtime/dotnet-legacy.d.ts:305]: Let's move it to `cwraps_internal` and `INTERNAL` please. No `.d.ts` at all.

**Comment by lambdageek** [comment]
Proposed Blazor change is https://github.com/dotnet/aspnetcore/pull/46909

File: src/mono/wasm/runtime/dotnet-legacy.d.ts (hunk 1 above)
Comment (pavelsavara) [reply src/mono/wasm/runtime/dotnet-legacy.d.ts]: > What's the difference between export_internal and cwraps_internal? Does it matter that these aren't C wraps - they're normal TS functions.

No it doesn't matter.

**Review by pavelsavara — APPROVED** [review]
ts changes LGTM

**Comment by naricc** [comment]
Changes look good to me, I was just wondering what the use case is; would it make more sense to implement [GC.TryStartNoGCRegion](https://learn.microsoft.com/en-us/dotnet/api/system.gc.trystartnogcregion?view=net-7.0) instead?

**Review by naricc — APPROVED** [review]

**Comment by BrzVlad** [comment]
Why does the blazor rendering code need to lock on gc ?

**Comment by lambdageek** [comment]
> Why does the blazor rendering code need to lock on gc ?

So the way blazor rendering works is that on the C# side they build up a managed object that describes the DOM changes that they want to apply, and then on the JS side they run something that crawls over this object and applies the DOM mutations.  They use unsafe `MonoObject*` accesses (ie, they get field offsets and then load pointers) to traverse the render tree.  The actual JS rendering step runs in a single browser event iteration and is purely JS code - it doesn't invoke any managed code.

In single-threaded wasm this is GC safe, because there's a single thread and it's currently busy doing JS work.
In multi-threaded wasm, there could be background tasks that might trigger a GC and cause the render tree to move.

In the short term it's not possible to change Blazor to use GC handles or anything like that.  So our plan is:
1. give them a way to globally block any GC
2. rewrite their rendering pass to use a GC-safe API (for example something like #69458)

**Comment by lambdageek** [comment]
@BrzVlad any objections?

**Review by BrzVlad — APPROVED** [review]

**Pull request merged**